File: src/main/java/net/greenjab/fixedminecraft/registry/item/map_book/MapBookIdCountsState.java
package net.greenjab.fixedminecraft.registry.item.map_book;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.greenjab.fixedminecraft.FixedMinecraft;
import net.minecraft.util.datafix.DataFixTypes;
import net.minecraft.world.level.saveddata.SavedData;
import net.minecraft.world.level.saveddata.SavedDataType;

public class MapBookIdCountsState extends SavedData {
    public static final Codec<MapBookIdCountsState> CODEC = RecordCodecBuilder.create(
             instance -> instance.group(
                     Codec.INT.optionalFieldOf("fixedminecraft:map_book", -1).forGetter( state -> state.nextMapBookId))
                    .apply(instance, MapBookIdCountsState::new)
    );

    int nextMapBookId;
    public MapBookIdCountsState() {
        nextMapBookId = -1;
    }

    public MapBookIdCountsState(int nextMapBookId) {
        this.nextMapBookId = nextMapBookId;
    }

    public int get() {
        nextMapBookId++;
        this.setDirty();
        return nextMapBookId;
    }
    public static final SavedDataType<MapBookIdCountsState> persistentStateType = new SavedDataType<>(
            FixedMinecraft.id("map_book/last_id"), MapBookIdCountsState::new, CODEC, DataFixTypes.SAVED_DATA_MAP_INDEX
    );

}

File: src/main/kotlin/net/greenjab/fixedminecraft/registry/item/map_book/MapBookIdCountsState.kt
package net.greenjab.fixedminecraft.registry.item.map_book

import it.unimi.dsi.fastutil.objects.Object2IntMap
import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap
import net.minecraft.datafixer.DataFixTypes
import net.minecraft.nbt.NbtCompound
import net.minecraft.registry.RegistryWrapper.WrapperLookup
import net.minecraft.world.PersistentState


class MapBookIdCountsState : PersistentState() {
    private val idCounts: Object2IntMap<String> = Object2IntOpenHashMap()

    init {
        idCounts.defaultReturnValue(-1)
    }

    override fun writeNbt(nbt: NbtCompound, wrapperLookup: WrapperLookup): NbtCompound {
        for (entry in idCounts.object2IntEntrySet()) {
            nbt.putInt(entry.key, entry.intValue)
        }

        return nbt
    }

    val nextMapBookId: Int
        get() {
            val i = idCounts.getInt("fixedminecraft:map_book") + 1
            idCounts.put("fixedminecraft:map_book", i)
            this.markDirty()
            return i
        }

    companion object {
        const val IDCOUNTS_KEY: String = "fixedminecraft_idcounts"

        val persistentStateType: Type<MapBookIdCountsState>
            get() = Type({ MapBookIdCountsState() },
                { nbt: NbtCompound, registryLookup: WrapperLookup? ->
                    fromNbt(
                        nbt,
                        registryLookup
                    )
                }, DataFixTypes.SAVED_DATA_MAP_INDEX
            )

        fun fromNbt(nbt: NbtCompound, registryLookup: WrapperLookup?): MapBookIdCountsState {
            val idCountsState = MapBookIdCountsState()

            for (string in nbt.keys) {
                if (nbt.contains(string, 99)) {
                    idCountsState.idCounts.put(string, nbt.getInt(string))
                }
            }

            return idCountsState
        }
    }
}
